tsdb/engine/tsm1/cache.go

```diff
@@ -138,6 +139,14 @@ func (c *Cache) Snapshot() *Cache {
 	snapshot := NewCache(c.maxSize)
 	snapshot.store = c.store
 	snapshot.size = c.size
+	snapshot.dirty = make(map[string]*entry)
```

Comment:
The mistake in my first attempt to fix this issue is that I was taking the wrong lock when taking a partial copy and to do the final snapshot - I was taking the snapshot lock, rather than the engine' caches lock. This revision only ever uses the engine cache's lock.